### PR TITLE
fix 9232

### LINKED ISSUE #9232: httpfs: IO Error: Connection error for HTTP HEAD when read_json_auto from S3 with wildcard
### What happens?

**Failure** on `read_json_auto` from S3 with **DuckDb 0.9.0 Preview** release 

`Reading large amount (~15000) of JSON files on S3 fails with error `Error: Invalid Error: IO Error: Connection error for HTTP HEAD to...`

Error Log for *0.9.0*
 
```
➜  bin duckdb-bin/0.9.0/duckdb
v0.9.0 0d84ccf478
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D INSTALL httpfs;
D LOAD httpfs;
D INSTALL aws;
D LOAD aws;
D INSTALL json;
D LOAD json;
D call load_aws_credentials('virt');
D from duckdb_settings() where name like '%s3%';
┌──────────────────────┬─────────────────────────────────────────────────┬─────────────────────────────────────────────────────────────────────┬────────────┐
│         name         │                      value                      │                             description                             │ input_type │
│       varchar        │                     varchar                     │                               varchar                               │  varchar   │
├──────────────────────┼─────────────────────────────────────────────────┼─────────────────────────────────────────────────────────────────────┼────────────┤
│ s3_uploader_thread…  │ 50                                              │ S3 Uploader global thread limit (default 50)                        │ UBIGINT    │
│ s3_uploader_max_fi…  │ 800GB                                           │ S3 Uploader max filesize (between 50GB and 5TB, default 800GB)      │ VARCHAR    │
│ s3_url_compatibili…  │ 0                                               │ Disable Globs and Query Parameters on S3 urls                       │ BOOLEAN    │
│ s3_use_ssl           │ 1                                               │ S3 use SSL (default true)                                           │ BOOLEAN    │
│ s3_uploader_max_pa…  │ 10000                                           │ S3 Uploader max parts per file (between 1 and 10000, default 10000) │ UBIGINT    │
│ s3_secret_access_key │ LGND9...............................8VW2        │ S3 Access Key                                                       │ VARCHAR    │
│ s3_session_token     │ IQoJb3JpZ2luX..............hc3QtMSJIMEYCIQDsg…  │ S3 Session Token                                                    │ VARCHAR    │
│ s3_url_style         │ vhost                                           │ S3 url style ('vhost' (default) or 'path')                          │ VARCHAR    │
│ s3_endpoint          │ s3.amazonaws.com                                │ S3 Endpoint (default 's3.amazonaws.com')                            │ VARCHAR    │
│ s3_access_key_id     │ ASIA............KAYK                            │ S3 Access Key ID                                                    │ VARCHAR    │
│ s3_region            │ us-east-1                                       │ S3 Region                                                           │ VARCHAR    │
├──────────────────────┴─────────────────────────────────────────────────┴─────────────────────────────────────────────────────────────────────┴────────────┤
│ 11 rows                                                                                                                                         4 columns │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
D create table customer_steps_prd as select "activities-steps"[1]."dateTime" as record_date, "activities-steps"[1]."value" as steps_taken, split_part(filename, '/', -2)::INTEGER as participant_id from read_json_auto(['s3://blah-blah-blah/bruh-bruh-bruh/INTRADAY_STEPS/NA/ACTIVITY_MEASUREMENTS/9*/*.JSON', 's3://blah-blah-blah/bruh-bruh-bruh/INTRADAY_STEPS/NA/ACTIVITY_MEASUREMENTS/8*/*.JSON'],filename=true);

Error: Invalid Error: IO Error: Connection error for HTTP HEAD to 'https://blah-blah-blah.s3.amazonaws.com/bruh-bruh-bruh/ACTIVITY_MEASUREMENTS/800304/2023-09-27-00.00.00_2023-09-27-23.59.59-4da5f0dfe543eda6497e575ac6713c5d.JSON'
D show tables;
┌─────────┐
│  name   │
│ varchar │
├─────────┤
│ 0 rows  │
└─────────┘

```

Same loading took a long time but finished successfully on **DuckDB `0.8.1`**

```
➜  bin duckdb-bin/0.8.1/duckdb
v0.8.1 6536a77232
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D INSTALL httpfs;
D LOAD httpfs;
D INSTALL json;
D LOAD json;
D
D
D
D from duckdb_settings() where name like '%s3%';
┌──────────────────────┬─────────────────────────────────────────────────┬─────────────────────────────────────────────────────────────────────┬────────────┐
│         name         │                      value                      │                             description                             │ input_type │
│       varchar        │                     varchar                     │                               varchar                               │  varchar   │
├──────────────────────┼─────────────────────────────────────────────────┼─────────────────────────────────────────────────────────────────────┼────────────┤
│ s3_uploader_thread…  │ 50                                              │ S3 Uploader global thread limit (default 50)                        │ UBIGINT    │
│ s3_uploader_max_fi…  │ 800GB                                           │ S3 Uploader max filesize (between 50GB and 5TB, default 800GB)      │ VARCHAR    │
│ s3_url_compatibili…  │ 0                                               │ Disable Globs and Query Parameters on S3 urls                       │ BOOLEAN    │
│ s3_use_ssl           │ 1                                               │ S3 use SSL (default true)                                           │ BOOLEAN    │
│ s3_uploader_max_pa…  │ 10000                                           │ S3 Uploader max parts per file (between 1 and 10000, default 10000) │ UBIGINT    │
│ s3_secret_access_key │ LGND957..........................6ll8VW2        │ S3 Access Key                                                       │ VARCHAR    │
│ s3_session_token     │ IQoJb3J..........................SJIMEYCIQDsg…  │ S3 Session Token                                                    │ VARCHAR    │
│ s3_url_style         │ vhost                                           │ S3 url style ('vhost' (default) or 'path')                          │ VARCHAR    │
│ s3_endpoint          │ s3.amazonaws.com                                │ S3 Endpoint (default 's3.amazonaws.com')                            │ VARCHAR    │
│ s3_access_key_id     │ ASIA............KAYK                            │ S3 Access Key ID                                                    │ VARCHAR    │
│ s3_region            │ us-east-1                                       │ S3 Region                                                           │ VARCHAR    │
├──────────────────────┴─────────────────────────────────────────────────┴─────────────────────────────────────────────────────────────────────┴────────────┤
│ 11 rows                                                                                                                                         4 columns │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

####### STARTED AT 11:12 AM

D create table customer_steps_prd as select "activities-steps"[1]."dateTime" as record_date, "activities-steps"[1]."value" as steps_taken, split_part(filename, '/', -2)::INTEGER as participant_id from read_json_auto(['s3://blah-blah-blah/bruh-bruh-bruh/ACTIVITY_MEASUREMENTS/9*/*.JSON', 's3://blah-blah-blah/bruh-bruh-bruh/INTRADAY_STEPS/NA/ACTIVITY_MEASUREMENTS/8*/*.JSON'],filename=true);

100% ▕████████████████████████████████████████████████████████████▏
#######  FINISHED AT 11:38 AM
D show tables;
┌────────────────────┐
│        name        │
│      varchar       │
├────────────────────┤
│ customer_pills     │
└────────────────────┘
D describe customer_pills;
┌────────────────┬─────────────┬─────────┬─────────┬─────────┬───────┐
│  column_name   │ column_type │  null   │   key   │ default │ extra │
│    varchar     │   varchar   │ varchar │ varchar │ varchar │ int32 │
├────────────────┼─────────────┼─────────┼─────────┼─────────┼───────┤
│ record_date    │ DATE        │ YES     │         │         │       │
│ pills_taken    │ BIGINT      │ YES     │         │         │       │
│ participant_id │ INTEGER     │ YES     │         │         │       │
└────────────────┴─────────────┴─────────┴─────────┴─────────┴───────┘
D select count(*) from customer_pills;
┌──────────────┐
│ count_star() │
│    int64     │
├──────────────┤
│        14590 │
└──────────────┘
D

```

**Outcomes**

* `DuckDB 0.8.1`
  * Succeeded after 26 minutes to read ~15000 JSON files from S3 extensions (httpfs+json)

* `DuckDB 0.9.0`
  * Failed after 3 minutes and did not complete for **_~15000 JSON Files_** from S3 extensions (httpfs+aws+json)
  * Succeeded after 5 minutes to read **_320 JSON Files_** from S3 extensions (httpfs+aws+json)

### To Reproduce

1. Connect with S3 
2. Fetch data from many small JSON files via `read_json_auto`, currently failed with file count ~15000
3. Read all files with wildcard globbing and extract fields from it and use filepath field too extract name of file

### OS:

MacOs 14.0  Sonoma 

### DuckDB Version:

0.9.1

### DuckDB Client:

CLI

### Full Name:

Binoy Shah

### Affiliation:

Vibrent Health

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by samansmink**
Followup of https://github.com/duckdb/duckdb/pull/9074
fixes https://github.com/duckdb/duckdb/issues/9232

httpfs files were left open, which in combination with keep-alive tcp connections would run into the limit of max open connections.

The solution is is a bit hacky, but wanted to get this in before 0.9.2 @Mytherin

### CODE REVIEW

File: extension/json/buffered_json_reader.cpp [hunk 1]
```diff
@@ -23,7 +23,7 @@ bool JSONFileHandle::IsOpen() const {
 }
 
 void JSONFileHandle::Close() {
-	if (IsOpen() && file_handle->OnDiskFile()) {
+	if (IsOpen() && file_handle->file_system.GetName() != "PipeFileSystem") {
```
Comment (Mytherin) [inline extension/json/buffered_json_reader.cpp:26]: Maybe calling `IsPipe` is cleaner?

Comment (samansmink) [reply extension/json/buffered_json_reader.cpp]: yes definitely

**Review by Mytherin — APPROVED** [review]
Thanks, LGTM

**Comment by mohitreddy1996** [comment]
@samansmink did this make it to the 0.9.2 release? (https://github.com/duckdb/duckdb/releases - I don't see this PR mentioned in the changelog). 

We are seeing similar issues as mentioned in the issue linked.

**Comment by Mytherin** [comment]
This PR is part of v0.9.2 yes, perhaps you could open a new issue?

File: extension/json/buffered_json_reader.cpp [hunk 2]
```diff
@@ -23,7 +23,7 @@ bool JSONFileHandle::IsOpen() const {
 }
 
 void JSONFileHandle::Close() {
-	if (IsOpen() && file_handle->OnDiskFile()) {
+	if (IsOpen() && !file_handle->IsPipe()) {
```
Comment (dentiny) [inline extension/json/buffered_json_reader.cpp:26]: curious why do we skip closing the pipe file handle here?

**Pull request merged**